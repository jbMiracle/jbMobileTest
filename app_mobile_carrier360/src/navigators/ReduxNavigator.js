import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { addNavigationHelpers, NavigationActions } from 'react-navigation';
import { createReduxBoundAddListener } from 'react-navigation-redux-helpers';
import { BackHandler } from 'react-native';
import { connect } from 'react-redux';
import { AppNavigator } from './AppNavigator';

export class ReduxNavigator extends Component {
  componentDidMount() {
    BackHandler.addEventListener('hardwareBackPress', this.onBackPress);
  }

  componentWillUnmount() {
    BackHandler.removeEventListener('hardwareBackPress', this.onBackPress);
  }

  onBackPress = () => {
    const { dispatch, nav } = this.props;
    if (nav.index === 0) {
      return false;
    }
    dispatch(NavigationActions.back());
    return true;
  };

  render() {
    const { dispatch, nav } = this.props;
    const addListener = createReduxBoundAddListener('root');
    const navHelpers = addNavigationHelpers({
      dispatch,
      state: nav,
      addListener,
    });

    return <AppNavigator navigation={ navHelpers } />;
  }
}
const mapStateToProps = state => ({
  nav: state.nav,
});

ReduxNavigator.propTypes = {
  nav: PropTypes.shape({
    index: PropTypes.number,
    isTransitioning: PropTypes.bool,
    routes: PropTypes.arrayOf(PropTypes.object),
  }).isRequired,
  dispatch: PropTypes.func.isRequired,
};

export default connect(mapStateToProps)(ReduxNavigator);
