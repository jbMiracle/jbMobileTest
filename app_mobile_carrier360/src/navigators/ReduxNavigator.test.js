import React from 'react';
import { BackHandler } from 'react-native';
import { ReduxNavigator } from './ReduxNavigator';

describe('ReduxNavigator', () => {
  describe('back navigation', () => {
    it('will navigate back if there is a stack', () => {
      const nav = {
        index: 1,
      };
      const dispatch = jest.fn();
      const reduxNavigator = new ReduxNavigator({ nav, dispatch });
      const canGoBack = reduxNavigator.onBackPress();

      expect(dispatch).toMatchSnapshot();
      expect(canGoBack).toBe(true);
    });

    it('won\'t navigate back if there is not a stack', () => {
      const nav = {
        index: 0,
      };
      const dispatch = jest.fn();
      const reduxNavigator = new ReduxNavigator({ nav, dispatch });
      const canGoBack = reduxNavigator.onBackPress();

      expect(dispatch).not.toHaveBeenCalled();
      expect(canGoBack).toBe(false);
    });
  });

  describe('lifecycle functions', () => {
    beforeEach(() => {
      BackHandler.addEventListener = undefined;
      BackHandler.removeEventListener = undefined;
    });

    it('will add the back listener on mount', () => {
      BackHandler.addEventListener = jest.fn();
      const reduxNavigator = new ReduxNavigator({});
      reduxNavigator.componentDidMount();
      expect(BackHandler.addEventListener).toMatchSnapshot();
    });

    it('will remove the back listener on unmount', () => {
      BackHandler.removeEventListener = jest.fn();
      const reduxNavigator = new ReduxNavigator({});
      reduxNavigator.componentWillUnmount();
      expect(BackHandler.removeEventListener).toMatchSnapshot();
    });
  });
});
