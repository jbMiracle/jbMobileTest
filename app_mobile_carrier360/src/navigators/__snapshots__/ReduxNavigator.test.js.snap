// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReduxNavigator back navigation will navigate back if there is a stack 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "immediate": undefined,
        "key": undefined,
        "type": "Navigation/BACK",
      },
    ],
  ],
}
`;

exports[`ReduxNavigator lifecycle functions will add the back listener on mount 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "hardwareBackPress",
      [Function],
    ],
  ],
}
`;

exports[`ReduxNavigator lifecycle functions will remove the back listener on unmount 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "hardwareBackPress",
      [Function],
    ],
  ],
}
`;
